packages/mgt-chat/src/statefulClient/GraphNotificationUserClient.ts

```diff
@@ -224,72 +224,96 @@ export class GraphNotificationUserClient {
-              if ((e as { statusCode?: number }).statusCode === 404) {
-                log('Removing subscription from cache', subscription.id);
-                await this.subscriptionCache.deleteCachedSubscriptions(this.currentUserId, this.sessionId);
-                await this.subscribeToUserNotifications(this.currentUserId);
-
-                const emitter: ThreadEventEmitter | undefined = this.emitter;
-                emitter?.reconnected();
-              }
-            }
+            await this.renewSubscription(this.currentUserId, subscription);
+          }
+        } catch (renewalEx) {
+          isRenewalInError = true;
+          // this error indicates we are not able to successfully renew the subscription, so we should create a new one.
+          if ((renewalEx as { statusCode?: number }).statusCode === 404) {
+            log('Removing subscription from cache');
+            await this.subscriptionCache.deleteCachedSubscriptions(this.currentUserId, this.sessionId);
+            subscription = undefined;
           } else {
-            // create a connection to the web socket if one does not exist
-            if (!this.connection) await this.createSignalRConnection(subscription.notificationUrl);
+            // log and continue (we will try again later)
+            error(renewalEx);
           }
+        }
+      }
+
+      if (!subscription) {
+        subscription = await this.createSubscription(this.currentUserId);
+      }
 
-          // Expecting only one subscription per user
-          break;
+      // notificationUrl comes in the form of websockets:https://graph.microsoft.com/beta/subscriptions/notificationChannel/websockets/<Id>?groupid=<UserId>&sessionid=default
+      // if <Id> changes, we need to create a new connection
+      if (
+        !this.connection ||
+        this.connection.state !== HubConnectionState.Connected ||
+        this.lastNotificationUrl !== subscription.notificationUrl
+      ) {
+        if (subscription.notificationUrl) {
+          this.lastNotificationUrl = subscription.notificationUrl;
+          await this.createSignalRConnection(subscription.notificationUrl);
+        } else {
+          // this should never happen.
+          throw new Error('Subscription notificationUrl is invalid.');
         }
       }
     } catch (e) {
+      isRenewalInError = true;
+      // log and continue (we will try again later)
       error(e);
     }
 
+    const isConnected = !isRenewalInError && this.connection?.state === HubConnectionState.Connected;
+    if (this.wasConnected !== isConnected) {
+      this.wasConnected = isConnected;
+      const emitter: ThreadEventEmitter | undefined = this.emitter;
+      if (isConnected) {
+        emitter?.connected();
+      } else {
+        emitter?.disconnected(false);
+      }
+    }
+
     this.isRewnewalInProgress = false;
     this.renewalInterval = this.timer.setTimeout(this.renewalSync, appSettings.renewalTimerInterval * 1000);
```

Comment:
I think this is fine, however, the PG may insist on exponential backoff, and if they do, that is easy enough to do here.

Reply:
Does the renewal every 10 seconds not cover retry?

Reply:
It does retry, my point on exponential backoff is that they may think 3 seconds (or even 10 seconds) is too spammy. Again, I would be fine without it, so I am not suggesting a change, just saying this may be something we have to revisit after PG reviews.